src/providers/core_provider/mod.rs

```diff
@@ -26,37 +27,27 @@ const SUPPORTED_OPCODES: [Opcode; 3] = [Opcode::ListProviders, Opcode::ListOpcod
 pub struct CoreProvider {
     wire_protocol_version_min: u8,
     wire_protocol_version_maj: u8,
-    providers: Vec<ProviderInfo>,
+    providers: HashMap<ProviderInfo, HashSet<Opcode>>,
```

Comment:
It might be a bit more efficient/simpler to add a `providers_opcodes: HashMap<ProviderID, HashSet<Opcode>>` and rename the current `providers` field `providers_info`. That way you do not have to iterate through all the providers and the logic is a bit more simple in `list_opcodes`.

Reply:
Indeed, but you'd have duplicate data (the provider ID). I'll try and figure out why we don't include `opcodes` in the `ProviderInfo` in the first place.

Reply:
Done